cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -560,6 +560,87 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
+                            // calculations.
+                            //
+                            //The sequence is as follows where A=lhs and B=rhs cl operands:
+                            //
+                            // A' = A
+                            // A'' = A
+                            // A = Al * Bl
+                            // A' = A' -> 32
+                            // A' = Ah' * Bl
+                            // A' = A' <- 32
+                            // A = A + A'
+                            // B = B -> 32
+                            // B = Bh * Al''
+                            // B = B <- 32
+                            // A = A + B
+
+                            // A' = A
+                            // A'' = A
+                            let lhs = input_to_reg(ctx, inputs[1]);
+                            let lhs_tmp1 = ctx.alloc_tmp(RegClass::V128, types::I64X2);
+                            let lhs_tmp2 = ctx.alloc_tmp(RegClass::V128, types::I64X2);
+                            ctx.emit(Inst::gen_move(lhs_tmp1, lhs, ty));
+                            ctx.emit(Inst::gen_move(lhs_tmp2, lhs, ty));
+
+                            // A = Al * Bl
+                            let rhs = input_to_reg_mem(ctx, inputs[0]);
+                            let dst = output_to_reg(ctx, outputs[0]);
+                            ctx.emit(Inst::gen_move(dst, lhs, ty));
+                            ctx.emit(Inst::xmm_rm_r(SseOpcode::Pmuludq, rhs.clone(), dst));
+
+                            // A' = A' -> 32
+                            // A' = Ah' * Bl
+                            // A' = A' <- 32
+                            // A = A + A'
+                            ctx.emit(Inst::xmm_rmi_reg(
+                                SseOpcode::Psrlq,
+                                RegMemImm::imm(32),
+                                lhs_tmp1,
+                            ));
+                            ctx.emit(Inst::xmm_rm_r(SseOpcode::Pmuludq, rhs.clone(), lhs_tmp1));
+                            ctx.emit(Inst::xmm_rmi_reg(
+                                SseOpcode::Psllq,
+                                RegMemImm::imm(32),
+                                lhs_tmp1,
+                            ));
+                            ctx.emit(Inst::xmm_rm_r(
+                                SseOpcode::Paddq,
+                                RegMem::reg(lhs_tmp1.to_reg()),
+                                dst,
+                            ));
+
+                            // B = B -> 32
+                            // B = Bh * Al''
+                            // B = B <- 32
+                            // A = A + B
+                            ctx.emit(Inst::xmm_rmi_reg(
+                                SseOpcode::Psrlq,
+                                RegMemImm::imm(32),
+                                Writable::from_reg(rhs.to_reg().unwrap()),
```

Comment:
It is an error to reuse an input register and write into it; the RHS might be reused after this particular CLIF instruction, and there's no indication to regalloc that it ought to be preserve it. Instead, we need to copy RHS into a temporary, and use that, here. Note the vreg copy might not become a copy machine instruction, since regalloc may coalesce those two vregs.

Also, `rhs` is a RegMem, so it can be either a register or memory, and here the `to_reg().unwrap()` assumes it's in a register. When defining `rhs`, you need to use `input_to_reg` and not `input_to_reg_mem` to reflect that you want it to be in a register.

Reply:
Hi @bnjbvr First of all note, in the latest push yesterday I switched which input referred to the lhs and which referred to the rhs as that was not consistent/correct. For this specific comment what is now the lhs needed the input to be input_to_reg_mem because of the Pmuludq instructions. Specifically https://github.com/bytecodealliance/wasmtime/blob/c2af20b113e509a603db78740cb146ea3fb2246b/cranelift/codegen/src/isa/x64/lower.rs#L599 and https://github.com/bytecodealliance/wasmtime/blob/c2af20b113e509a603db78740cb146ea3fb2246b/cranelift/codegen/src/isa/x64/lower.rs#L610. Am I missing understanding something here?

Reply:
`input_to_reg_mem` means that the input will be put into a register *or* a memory operand, and we don't know which one. So we can't use `to_reg().unwrap()` on the result of an `input_to_reg_mem` call, because the operand might be in memory, and `to_reg()` returns `None` in this case. Instead, you want to use `input_to_reg`, which will always return a register operand, and create a `RegMem::reg(...)` operand when you need to use it for an instruction that requires a `RegMem`. Does it help?

Reply:
Ok .. And I see you've renamed this :-). So I am still a little uncertain about the reuse of input registers .. is it at any time to use them as a temporary output? I've reworked the algorithm to factor out that shift and modified some things without using too many temps but it's not clear to me if what I have in this update I am about to push is the optimal usage.